File: test/com/developerphil/adbidea/compatibility/BackwardCompatibleGetterTest.java
package com.developerphil.adbidea.compatibility;

import com.developerphil.adbidea.test.Holder;
import org.joor.ReflectException;
import org.junit.Test;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;

public class BackwardCompatibleGetterTest {


    @Test
    public void onlyCallCurrentImplementationWhenItIsValid() throws Exception {
        final Holder<Boolean> currentImplementation = new Holder<Boolean>(false);

        new BackwardCompatibleGetter<Boolean>() {
            @Override
            protected Boolean getCurrentImplementation() throws Throwable {
                currentImplementation.set(true);
                return true;
            }

            @Override
            protected Boolean getPreviousImplementation() {
                fail("should not be called");
                return true;
            }
        }.get();

        assertThat(currentImplementation.get(), is(true));

    }

    @Test
    public void callPreviousImplementationWhenCurrentThrowsErrors() throws Exception {
        expectPreviousImplementationIsCalledFor(new ClassNotFoundException());
        expectPreviousImplementationIsCalledFor(new NoSuchMethodException());
        expectPreviousImplementationIsCalledFor(new NoSuchFieldException());
        expectPreviousImplementationIsCalledFor(new LinkageError());
        expectPreviousImplementationIsCalledFor(new ReflectException());
    }


    @Test(expected = RuntimeException.class)
    public void throwExceptionsWhenTheyAreNotRelatedToBackwardCompatibility() throws Exception {
        new BackwardCompatibleGetter<Boolean>() {
            @Override
            protected Boolean getCurrentImplementation() throws Throwable {
                throw new RuntimeException("exception!");
            }

            @Override
            protected Boolean getPreviousImplementation() {
                fail("should not be called");
                return true;
            }
        }.get();
    }


    private static void expectPreviousImplementationIsCalledFor(final Throwable throwable) throws Exception {
        final Holder<Boolean> calledPreviousImplementation = new Holder<Boolean>();
        new BackwardCompatibleGetter<Boolean>() {
            @Override
            protected Boolean getCurrentImplementation() throws Throwable {
                throw throwable;
            }

            @Override
            protected Boolean getPreviousImplementation() {
                calledPreviousImplementation.set(true);
                return null;
            }
        }.get();

        assertThat(calledPreviousImplementation.get(), is(true));
    }

}

File: src/test/kotlin/com/developerphil/adbidea/compatibility/BackwardCompatibleGetterTest.kt
package com.developerphil.adbidea.compatibility

import com.google.common.truth.Truth.assertThat
import org.joor.ReflectException
import org.junit.Assert.fail
import org.junit.Test

class BackwardCompatibleGetterTest {
    @Test
    fun onlyCallCurrentImplementationWhenItIsValid() {
        var value = false
        object : BackwardCompatibleGetter<Boolean>() {
            override fun getCurrentImplementation(): Boolean {
                value = true
                return true
            }

            override fun getPreviousImplementation(): Boolean {
                fail("should not be called")
                return true
            }
        }.get()
        assertThat(value).isTrue()
    }

    @Test
    fun callPreviousImplementationWhenCurrentThrowsErrors() {
        expectPreviousImplementationIsCalledFor(ClassNotFoundException())
        expectPreviousImplementationIsCalledFor(NoSuchMethodException())
        expectPreviousImplementationIsCalledFor(NoSuchFieldException())
        expectPreviousImplementationIsCalledFor(LinkageError())
        expectPreviousImplementationIsCalledFor(ReflectException())
    }

    @Test(expected = RuntimeException::class)
    fun throwExceptionsWhenTheyAreNotRelatedToBackwardCompatibility() {
        object : BackwardCompatibleGetter<Boolean>() {
            override fun getCurrentImplementation(): Boolean {
                throw RuntimeException("exception!")
            }

            override fun getPreviousImplementation(): Boolean {
                fail("should not be called")
                return false
            }
        }.get()
    }

    private fun expectPreviousImplementationIsCalledFor(throwable: Throwable) {
        var value = false
        object : BackwardCompatibleGetter<Boolean>() {
            override fun getCurrentImplementation(): Boolean {
                throw throwable
            }

            override fun getPreviousImplementation(): Boolean {
                value = true
                return true
            }
        }.get()

        assertThat(value).isTrue()
    }
}